examples/inference/inpainting.py

```diff
@@ -0,0 +1,182 @@
+import inspect
+from typing import List, Optional, Union
+
+import numpy as np
+import torch
+
+import PIL
+from diffusers import AutoencoderKL, DDIMScheduler, DiffusionPipeline, PNDMScheduler, UNet2DConditionModel
+from diffusers.pipelines.stable_diffusion import StableDiffusionSafetyChecker
+from tqdm.auto import tqdm
+from transformers import CLIPFeatureExtractor, CLIPTextModel, CLIPTokenizer
+
+
+def preprocess(image):
+    w, h = image.size
+    w, h = map(lambda x: x - x % 32, (w, h))  # resize to integer multiple of 32
+    image = image.resize((w, h), resample=PIL.Image.LANCZOS)
+    image = np.array(image).astype(np.float32) / 255.0
+    image = image[None].transpose(0, 3, 1, 2)
+    image = torch.from_numpy(image)
+    return 2.0 * image - 1.0
+
+def preprocess_mask(mask):
+    mask=mask.convert("L")
+    mask = mask.resize((64,64), resample=PIL.Image.LANCZOS)
```

Comment:
I think the mask needs to be shape (64,64).  Won't this fail if the original mask size isn't (512,512)?

Reply:
Stable diffusion can be used with images variable size, for example `512x512`, `728x512` etc. This will allow the pipeline to handle custom height, width. 

We should add a check here that image and mask size are equal and are divisible by 8. * because the VAE is f8, it reduces the images by a factor of 8.

Reply:
hmm... is there something in my pipeline I need to change for images that aren't 512x512?  When I try on an image that is 512x544 I get the error

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [26], in <cell line: 1>()
      3 mask_img=Image.open(mask_img_path).convert('RGB')
      4 for i in range(n):
----> 6   imgOut=infer(prompt,init_img,mask_img,prompt_strength)
      7   imgPath=os.path.join(sample_path, "%05d.png"%base_count)
      8   base_count+=1

Input In [25], in infer(prompt, init_image, mask_image, strength)
      4     mask_blur=mask_image.filter(ImageFilter.GaussianBlur(radius = 8))
      5     with autocast("cuda"):
----> 6         images = pipeimg([prompt] * num_samples, 
      7                          init_image=init_image,
      8                          mask_image=mask_image,
      9                          strength=strength, 
     10                          num_inference_steps=200,
     11                          guidance_scale=7.5)["sample"]
     12         images= [Image.composite(init_image,image,mask_image) for image in images]
     13 else: 

File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\autograd\grad_mode.py:27, in _DecoratorContextManager.__call__.<locals>.decorate_context(*args, **kwargs)
     24 @functools.wraps(func)
     25 def decorate_context(*args, **kwargs):
     26     with self.clone():
---> 27         return func(*args, **kwargs)

File E:\img\diffusers\examples\inference\inpainting.py:164, in StableDiffusionInpaintingPipeline.__call__(self, prompt, init_image, mask_image, strength, num_inference_steps, guidance_scale, eta, generator, output_type)
    161 latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
    163 # predict the noise residual
--> 164 noise_pred = self.unet(latent_model_input, t, encoder_hidden_states=text_embeddings)["sample"]
    166 # perform guidance
    167 if do_classifier_free_guidance:

File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\nn\modules\module.py:1110, in Module._call_impl(self, *input, **kwargs)
   1106 # If we don't have any hooks, we want to skip the rest of the logic in
   1107 # this function, and just call forward.
   1108 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1109         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1110     return forward_call(*input, **kwargs)
   1111 # Do not call functions when jit is used
   1112 full_backward_hooks, non_full_backward_hooks = [], []

File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\diffusers\models\unet_2d_condition.py:168, in UNet2DConditionModel.forward(self, sample, timestep, encoder_hidden_states)
    165 down_block_res_samples = down_block_res_samples[: -len(upsample_block.resnets)]
    167 if hasattr(upsample_block, "attentions") and upsample_block.attentions is not None:
--> 168     sample = upsample_block(
    169         hidden_states=sample,
    170         temb=emb,
    171         res_hidden_states_tuple=res_samples,
    172         encoder_hidden_states=encoder_hidden_states,
    173     )
    174 else:
    175     sample = upsample_block(hidden_states=sample, temb=emb, res_hidden_states_tuple=res_samples)

File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\nn\modules\module.py:1110, in Module._call_impl(self, *input, **kwargs)
   1106 # If we don't have any hooks, we want to skip the rest of the logic in
   1107 # this function, and just call forward.
   1108 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1109         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1110     return forward_call(*input, **kwargs)
   1111 # Do not call functions when jit is used
   1112 full_backward_hooks, non_full_backward_hooks = [], []

File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\diffusers\models\unet_blocks.py:1034, in CrossAttnUpBlock2D.forward(self, hidden_states, res_hidden_states_tuple, temb, encoder_hidden_states)
   1032 res_hidden_states = res_hidden_states_tuple[-1]
   1033 res_hidden_states_tuple = res_hidden_states_tuple[:-1]
-> 1034 hidden_states = torch.cat([hidden_states, res_hidden_states], dim=1)
   1036 hidden_states = resnet(hidden_states, temb)
   1037 hidden_states = attn(hidden_states, context=encoder_hidden_states)

RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 18 but got size 17 for tensor number 1 in the list.
```

(I get the same error when running the img2img pipeline with an init_image that is 512x544)

Reply:
- make sure both images and mask image are of similer shape and sizes are divisible by 8
- when creating the mask resize it to `(mask.size[0]//8,  mask.size[1]//8)` rather than `(64, 64)`.
WIll also try this.